Add documentation for the Azure DevOps extension and Issue Tracker (#466)
OctopusDeploy/docs: docs/deployment-process/issue-tracking/azure-devops.md

---
title: Azure DevOps Work Item Tracking Integration
description: Configure Azure DevOps work item tracking with Octopus.
position: 40
---

**Octopus 2019.7.3** introduced support to integrate Octopus with Azure DevOps work items. The integration includes the ability to:
- Automatically add links to Azure DevOps work items from releases and deployments in Octopus.

> - Retrieve release notes from Azure DevOps work item comments to assist in automating release note generation.

Reviewer: Suggesting this edit to make sure it doesn't alter the meaning.
```suggestion
- Retrieve release notes from Azure DevOps work item comments for automatic release note generation.
```

Author: Ok done, and also applied to GitHub page.

## How Azure DevOps Integration Works

![Octopus Azure DevOps integration - how it works diagram](images/octo-azure-devops-how-it-works.png "width=500")

> 1. Associate code changes with their relevant work items in any of the following ways:

Reviewer 2: Do you think this warrants a breakout section like the GitHub page did for commits? To clearly explain the different options w/ screenshots?

Author: I'd be concerned about disrupting flow: I want to at least lay out the user's set of options in the main flow here.

I wouldn't be against adding a section if further explanation is needed, but I can only imagine an animated screenshot involving "locate, zoom, click, click, click" because these interactions wouldn't convey well in a static screenshot. Maybe something to plan towards?

    - Edit a pull request in Azure DevOps, and use the **Work Items** panel to select a work item.
    - Edit a work item in Azure DevOps, and use the **Development** panel to add a pull request link (before build), or a commit link, or a build link.

>     - When you commit code, include "#" followed by a valid work item ID in the commit message. *Note: Requires first enabling the repository setting "Automatically create links for work items mentioned in a commit comment" under Project Settings (Repositories).*

Reviewer 2: This looks a bit messy. It might be better to add the note on a new line? @wordlee can comment on structure better though. :)

    - Unless the Azure DevOps instance is public, you'll need to supply a **Personal Access Token**, created in the Azure DevOps User Settings (under Personal access tokens), with authorization to read scopes **Build** and **Work items**.

    - Ensure the **Is Enabled** property is set as well.

2. Configure the Release Note Options _(optional)_.

    - **Release Note Prefix**: If specified, Octopus will look for a work item comment that starts with the given prefix text and use whatever text appears after the prefix as the release note, which will be available in the [build information](/docs/packaging-applications/build-servers/index.md#build-information) as the work item's description. If no comment is found with the prefix then Octopus will default back to using the title for that work item.

    For example, a prefix of `Release note:` can be used to identify a customer friendly work item title vs a technical feature or bug fix title.

When configured, this integration will retrieve Azure DevOps work item details and add details to your releases and deployments.

## Troubleshooting

> 

Reviewer 2: Probably good to add an intro note here. `This section outlines common issues and how to resolve them.`

Reviewer: I've reworded/restructured this (commit pending). It's still wordy but the option name is quite long :) : 
When you commit code, if you've enabled the repository setting: **Automatically create links for work items mentioned in a commit comment** under Project Settings (Repositories), you can include `#` followed by a valid work item ID in the commit message.

### Error message: "Unable to obtain some work item data."

The rest of the message will have more information about what went wrong, but this typically indicates a failure to do one of the following:

> * contact Azure DevOps over the network

Reviewer 2: Probably good to capitalise each point. i.e. `C`ontact Azure DevOps over the network

* authenticate using the specified Personal Access Token to read scopes **Build** and **Work items**
* retrieve build and work item data, and work item comments
* interpret the data

> 

Reviewer 2: Would it make sense to offer a suggestion here. Something like `It should be straightforward to resolve most of these issues. For example, if there is a problem retrieving build and work item details, ensure that your Personal Access Token has appropriate access to the right scopes as mentioned in the section [above](link).`

Author: Maybe, but how can I avoid sentences like "If it's your network that is broken, try fixing your network" ? It's one thing to speculate about what's wrong, like I am doing in this passage, but I feel like there's too little to go on to turn it into a recommendation.

Author: After more discussion with @wordlee I'm removing the section and will add suggestion to error messages where appropriate.

> If you can't resolve the problem and want to create a release, some options are:

Reviewer 2: Do we want to include this information? i.e. Turn it off? 😬

Author: Well, we don't want customers to accidentally create releases if these linkages are needing to be repaired. But we also don't want them to be unable to release if they can't repair.

Author: Should I keep it simple with one sentence? Which would be like this -
```markdown
If you can't resolve the problem and want to create a release, disable the Azure DevOps Issue Tracker extension under **{{Configuration,Settings,Azure DevOps Issue Tracker}}**.
```

Author: I can just remove the bit about not pushing metadata, since they probably have if they've hit this this point.

Reviewer: You could add to line 75, `if you ever need to disable the Azure DevOps Issue Tracker extension you can do so under **{{Configuration,Settings,Azure DevOps Issue Tracker}}**.`

Reviewer: That avoids explicitly saying turn it off, but it provides the information if they need it.

Author: Will do.
